Log Normal Sampler

SaitejaUtpala opened the pull request:
This is initiation of new experimental sampling module with  Log Normal Sampler for SPDmanifold and Euclidean Space

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+import geomstats.geometry.spd_matrices as spd
```
Yayy, the sampling module! 🎉 

This file needs an explanatory docstring at the top.

Imports should be in alphabetical order.

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+import geomstats.geometry.spd_matrices as spd
+import geomstats.backend as gs
+
+class LogNormal:
```
That's awesome!! 🎉

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+import geomstats.backend as gs
+
+class LogNormal:
+    """ LogNormal Sampler 
```
NIT: Could you just review our conventions for docstrings? For example, there needs to be a period at the end of end function description, and parameter description (below), as well as a space before the ":" in the parameter descriptions. Thank you! 🙏

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+class LogNormal:
+    """ LogNormal Sampler 
+
+    Parameters:
```
No ":" at the end of Parameters

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+        Mean of the Distribution
+    cov: array-like, shape=[..., dim]
+        Covariance of the Distribution. Should be Positive Semi Definite
+    Returns:
```
One blank line before Returns, no : at the end of Returns

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+
+    References:
+    ----------
+    Lognormal Distributions and Geometric Averages of Symmetric Positive Definite Matrices
```
Yesss for the reference! See here, for example, to see for how we cite papers in the code:
https://github.com/geomstats/geomstats/blob/909b77b18f2f4c1a8a61f661671412dfa165ee78/geomstats/geometry/discrete_curves.py#L90

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+    Lognormal Distributions and Geometric Averages of Symmetric Positive Definite Matrices
+    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5222531/
+    """
+    def __init__(self,manifold,mean,cov=None):
```
The code is not PEP8 compliant, e.g. there should be spaces after the ",", etc... Maybe DeepSource is catching this? Otherwise, it could be good to run flake8 to check.

ninamiolane reviewed (changes requested):
Very nice!! 🎉  @nguigs is that consistent with the idea you had for the sampling module?

ninamiolane reviewed (commented):
This also needs a test_ file.

nguigs commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,73 @@
+        _samples[:,i,i] = samples_euclidean[:,:self.n]
+        _samples[:,j,k] = off_diag
+        _samples[:,k,j] = off_diag
+        samples_spd = SPDmanifold.expm(_samples)
```
Maybe it should be discussed somewhere that 'Log' in `LogNormal` does not necessarily refer to Riemannian Log. For SPD, it does with the Log-Euclidean metric if I understand correctly, and asymptotically for the Affine-Invariant metric?

SaitejaUtpala replied:
> Maybe it should be discussed somewhere that 'Log' in `LogNormal` does not necessarily refer to Riemannian Log. For SPD, it does with the Log-Euclidean metric if I understand correctly, and asymptotically for the Affine-Invariant metric?

Done

nguigs reviewed (commented):
Nice! Thanks for getting this module started!
- When adding a test, you'll face a `NotImplementedError` for the multivariate_normal in tensorflow. If you can add it to that backend it would be great!
- Do you think that the distribution from this paper could be added in the same module (in a future PR) or it it is too different as the Riemannian exp is used? http://proceedings.mlr.press/v97/nagano19a.html

SaitejaUtpala commented:
> Nice! Thanks for getting this module started!
> 
>     * When adding a test, you'll face a `NotImplementedError` for the multivariate_normal in tensorflow. If you can add it to that backend it would be great!
> 
>     * Do you think that the distribution from this paper could be added in the same module (in a future PR) or it it is too different as the Riemannian exp is used? http://proceedings.mlr.press/v97/nagano19a.html

Seems really interesting reference. 

> Nice! Thanks for getting this module started!
> 
>     * When adding a test, you'll face a `NotImplementedError` for the multivariate_normal in tensorflow. If you can add it to that backend it would be great!
> 
>     * Do you think that the distribution from this paper could be added in the same module (in a future PR) or it it is too different as the Riemannian exp is used? http://proceedings.mlr.press/v97/nagano19a.html

seems really interesting method. have to see the paper more throroughly

nguigs commented:
> seems really interesting method. have to see the paper more throroughly

Ok, note that parallel transport and exp map is already implemented for the hyperboloid, so it should not be too much work. Also you can have a look at the current [spherical normal](https://github.com/geomstats/geomstats/blob/0040b7ef6f9b1e68f8c275d1b84c5567f17eec7a/geomstats/geometry/hypersphere.py#L434) that is quite similar I think (btw it will be moved to your new module)

SaitejaUtpala commented:
> > seems really interesting method. have to see the paper more throroughly
> 
> Ok, note that parallel transport and exp map is already implemented for the hyperboloid, so it should not be too much work. Also you can have a look at the current [spherical normal](https://github.com/geomstats/geomstats/blob/0040b7ef6f9b1e68f8c275d1b84c5567f17eec7a/geomstats/geometry/hypersphere.py#L434) that is quite similar I think (btw it will be moved to your new module)

Ok on the other note. Should we collect *all* the papers that are related to sampling on manifolds .  Because that will give good idea of how to design the module in better way and also know how much we have covered and what is left ?

ninamiolane commented:
> Ok on the other note. Should we collect _all_ the papers that are related to sampling on manifolds . Because that will give good idea of how to design the module in better way and also know how much we have covered and what is left ?

Yes, I think that would be great! The way sampling is done here could also be interesting in terms of module's design: https://github.com/geoopt/geoopt/tree/master/geoopt/samplers


And at some point, we could also have a list of the papers corresponding to all the methods (not only the sampling methods) that are implemented in the package! For example, as a web page in the documentation website, or in each of the READMEs of the corresponding geomstats modules?

nguigs commented:
> And at some point, we could also have a list of the papers corresponding to all the methods (not only the sampling methods) that are implemented in the package! For example, as a web page in the documentation website

Could this be done automatically from the references in the docstrings?

SaitejaUtpala commented:
RTM @ninamiolane  @nguigs 
following are changes

- make it work with tensorflow, and code is vectorized
- tests for auxillary backend function (@nguigs symmetric cls's `from_vector` and `to_vector` are different from what we want here)
- tests for log_normal distribution
- fixed linting and deep source issues
- changed doc-strings to meet `geomstats` convention

Also, backend is a complete mess; manually written functions are riddled with bugs.

SaitejaUtpala commented:
> > And at some point, we could also have a list of the papers corresponding to all the methods (not only the sampling methods) that are implemented in the package! For example, as a web page in the documentation website
> 
> Could this be done automatically from the references in the docstrings?

@nguigs yup, I can write a script that does this.

ninamiolane commented on geomstats/_backend/pytorch/__init__.py:
```diff
@@ -519,6 +519,10 @@ def equal(a, b, **kwargs):
 
 
+def diag_indices(*args, **kwargs):
+    return tuple(map(torch.from_numpy, _np.diag_indices(*args, **kwargs)))
```
We have been using the `torch.from_numpy` function a lot in this backend, ...but this is bad and was meant to be a hack: it might fail when it is used with pytorch tensors on GPUs (because numpy would not know how to compute on this).

To avoid this, could we get an implementation of diag_indices that is fully pytorch?

Many functions of the backends need a similar cleanup...

SaitejaUtpala replied:
So I working on other PR that updates pytorch version and clean the backend. They have added triu_indices (https://pytorch.org/docs/stable/generated/torch.triu_indices.html) and tril_indices (https://pytorch.org/docs/stable/generated/torch.tril_indices.html) in 1.8. Nothing has been added for diag_indices yet unfortunately. so I will remove diag_indices and find another way to get the result. And clean pytorch backend in the new PR.

ninamiolane commented on geomstats/_backend/pytorch/__init__.py:
```diff
@@ -787,3 +791,16 @@ def triu_to_vec(x, k=0):
+
+
+def mat_from_diag_triu_tril(diag, tri_upp, tri_low):
+    """build matrices (or batch of matrices) from diag and
```
NIT: The docstring needs to follow our docstring conventions

SaitejaUtpala replied:
> NIT: The docstring needs to follow our docstring conventions

Done

ninamiolane commented on geomstats/_backend/tensorflow/__init__.py:
```diff
@@ -748,6 +748,10 @@ def tril(m, k=0):
     return tf.linalg.band_part(m, -1, 0)
 
 
+def diag_indices(*args, **kwargs):
```
Same comment here, can we get a full tensorflow implementation instead?

SaitejaUtpala replied:
I have removed it. I am not using it anyway. Regarding cleaning, there are tensorflow operations based on Numpy API but it is still experimental (https://www.tensorflow.org/api_docs/python/tf/experimental/numpy). If this is present we would definitely need to migrate to it.

ninamiolane commented on geomstats/_backend/tensorflow/__init__.py:
```diff
@@ -787,3 +791,47 @@ def tile(x, multiples):
+
+
+def vec_to_triu(vec):
+    """Takes vec (batch of vecs) and forms
```
NIT: needs to respect our docstring conventions (Takes -> Take, only one line for the short description, the rest of the docstring goes in the longer description).

SaitejaUtpala replied:
Done

> NIT: needs to respect our docstring conventions (Takes -> Take, only one line for the short description, the rest of the docstring goes in the longer description).

Done

ninamiolane commented on geomstats/_backend/tensorflow/__init__.py:
```diff
@@ -787,3 +791,47 @@ def tile(x, multiples):
+
+def vec_to_triu(vec):
+    """Takes vec (batch of vecs) and forms
+    strictly upper traingular matrix (batch of mats) out of it"""
```
NIT: traingular -> triangular

SaitejaUtpala replied:
> NIT: traingular -> triangular

Done

ninamiolane commented on geomstats/_backend/tensorflow/__init__.py:
```diff
@@ -787,3 +791,47 @@ def tile(x, multiples):
+
+
+def vec_to_tril(vec):
+    """Takes vec ( batch of vec) and forms
```
NIT: Docstring convention

ninamiolane commented on geomstats/_backend/tensorflow/__init__.py:
```diff
@@ -787,3 +791,47 @@ def tile(x, multiples):
+    _triu = vec_to_triu(tri_upp)
+    _tril = vec_to_tril(tri_low)
+    triu_tril = _triu + _tril
+    mat = tf.linalg.set_diag(triu_tril, diag)
```
NIT: NIT: Docstring convention

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+class LogNormal:
+    """LogNormal Sampler on manifold of SPD Matrices and Euclidean Spaces.
+
+    (1)For Euclidean Spaces, if X is distributed as Normal(mean, cov), then
```
NIT: add a space after (1) and after (2)

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+
+    Parameters
+    ----------
+    manifold: Manifold obj, {Euclidean(n), SPDMatrices(n)}
```
NIT: `manifold : ` (space before and after :  )


same for below

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+    ----------
+    manifold: Manifold obj, {Euclidean(n), SPDMatrices(n)}
+        Manifold to sample over. Manifold should
+        be of instance of Euclidean or SPDMatrices
```
NIT: missing period at the end of parameter description

(also at other places below)

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+            shape=[dim, dim] if manifold is Euclidean Space
+            shape=[n*(n+1)/2, n*(n+1)/2] if manifold is SPD Manifold
+        Covariance of the Distribution.
+
```
No blank line here

ninamiolane commented on tests/tests_geomstats/test_backends.py:
```diff
@@ -1003,6 +1003,31 @@ def loss_torch(p):
         self.assertAllClose(result[0], value.detach())
         self.assertAllClose(result[1], grad)
 
```
Also test `vec_to_triu` and `vec_to_tril`?

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+
+    Examples
+    --------
+    import geomstats.backend as gs
```
Nice, very good idea to add an example!! Could you use this syntax? https://github.com/geomstats/geomstats/blob/eb1716ee3d8553123c779c23af65ecac5ec0aa08/geomstats/learning/online_kmeans.py#L144

This way, the example would better render on our documentation website.

ninamiolane commented on tests/tests_geomstats/test_lognormal.py:
```diff
@@ -0,0 +1,86 @@
+
+        expected = mean
+        result = fm
+        self.assertAllClose(result, expected, atol=5 * 1e-2)
```
Why is atol so low? would `gs.atol` work? (we ask for different precision depending on the backend, hence the `gs.atol`)

SaitejaUtpala replied:
Here we compute mean from drawn samples and see if it is close to mean of the distribution and so the answer is, we cannot expect more. It would essentially depend on sample efficiency of mean-estimator and all

ninamiolane commented on tests/tests_geomstats/test_lognormal.py:
```diff
@@ -0,0 +1,86 @@
+        result = fm
+        self.assertAllClose(result, expected, atol=5 * 1e-2)
+
+    def test_error_handling(self):
```
Nice!!

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+    cov: array-like,
+            shape=[dim, dim] if manifold is Euclidean Space
+            shape=[n*(n+1)/2, n*(n+1)/2] if manifold is SPD Manifold
+        Covariance of the Distribution.
```
NIT: Distribution -> distribution

SaitejaUtpala replied:
> NIT: Distribution -> distribution

Done

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+    from geomstats.geometry.spd_matrices import SPDMatrices
+    from geomstats.sampling.lognormal import LogNormal
+
+    mean = 2*gs.eye(3)
```
NIT: 2 * gs.eye(3)

(spaces around operator)

SaitejaUtpala replied:
> NIT: 2 * gs.eye(3)
> 
> (spaces around operator)

Done

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+    ----------
+    manifold: Manifold obj, {Euclidean(n), SPDMatrices(n)}
+        Manifold to sample over. Manifold should
+        be of instance of Euclidean or SPDMatrices
```
NIT: should be an instance of...

SaitejaUtpala replied:
> NIT: should be an instance of...

Done

ninamiolane commented on geomstats/sampling/lognormal.py:
```diff
@@ -0,0 +1,131 @@
+    (1)For Euclidean Spaces, if X is distributed as Normal(mean, cov), then
+    exp(X) is distributed as LogNormal(mean, cov).
+
+    (2)For SPDMatrices, LogNormal distribution is defined by embedding mean
```
For SPDMatrices, the LogNormal distribution is defined by transforming the mean -- an SPD matrix -- into a symmetric matrix through the Matrix logarithm, which gives the element "log-mean" that now belongs to a vector space. This log-mean and given cov are used to parametrize a Normal Distribution.

SaitejaUtpala replied:
> For SPDMatrices, the LogNormal distribution is defined by transforming the mean -- an SPD matrix -- into a symmetric matrix through the Matrix logarithm, which gives the element "log-mean" that now belongs to a vector space. This log-mean and given cov are used to parametrize a Normal Distribution.

Done.

ninamiolane reviewed (approved):
Great work!! Thank you so much for this addition. I have put many comments, but they are all minors. Good to merge on my side when these are addressed! Congrats!

SaitejaUtpala commented:
@ninamiolane . I have addressed review comments

ninamiolane reviewed (approved):
Awesome, thank you!! We have a sampling module 🎉  I let you decide if you want to implement the sampling from the distribution that @nguigs suggested.
